src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngAlgorithm.cs

```diff
@@ -0,0 +1,290 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.Contracts;
+
+namespace System.Security.Cryptography
+{
+    /// <summary>
+    ///     Utility class to strongly type algorithms used with CNG. Since all CNG APIs which require an
+    ///     algorithm name take the name as a string, we use this string wrapper class to specifically mark
+    ///     which parameters are expected to be algorithms.  We also provide a list of well known algorithm
+    ///     names, which helps Intellisense users find a set of good algorithm names to use.
+    /// </summary>
+    public sealed class CngAlgorithm : IEquatable<CngAlgorithm>
+    {
+        public CngAlgorithm(string algorithm)
+        {
+            Contract.Ensures(!string.IsNullOrEmpty(_algorithm));
+
+            if (algorithm == null)
+                throw new ArgumentNullException("algorithm");
+            if (algorithm.Length == 0)
+                throw new ArgumentException(SR.Format(SR.Cryptography_InvalidAlgorithmName, algorithm), "algorithm");
+
+            _algorithm = algorithm;
+        }
+
+        /// <summary>
+        ///     Name of the algorithm
+        /// </summary>
+        public string Algorithm
+        {
+            get
+            {
+                Contract.Ensures(!string.IsNullOrEmpty(Contract.Result<string>()));
+                return _algorithm;
+            }
+        }
+
+        public static bool operator ==(CngAlgorithm left, CngAlgorithm right)
+        {
+            if (object.ReferenceEquals(left, null))
+            {
+                return object.ReferenceEquals(right, null);
+            }
+
+            return left.Equals(right);
+        }
+
+        [Pure]
```

Comment:
If we're going to use [Pure], seems like we should use it consistently... why is != Pure but == isn't?